pkg/cmd/pr/shared/commentable.go

```diff
@@ -0,0 +1,163 @@
+	ConfirmSubmitSurvey   func() (bool, error)
+	OpenInBrowser         func(string) error
+	Interactive           bool
+	InputType             InputType
+	Body                  string
+}
+
+func CommentablePreRun(cmd *cobra.Command, opts *CommentableOptions) error {
+	inputFlags := 0
+	if cmd.Flags().Changed("body") {
+		opts.InputType = InputTypeInline
+		inputFlags++
+	}
+	if web, _ := cmd.Flags().GetBool("web"); web {
+		opts.InputType = InputTypeWeb
+		inputFlags++
+	}
+	if editor, _ := cmd.Flags().GetBool("editor"); editor {
+		opts.InputType = InputTypeEditor
+		inputFlags++
+	}
+
+	if inputFlags == 0 {
+		if !opts.IO.CanPrompt() {
+			return &cmdutil.FlagError{Err: errors.New("--body or --web required when not running interactively")}
+		}
+		opts.Interactive = true
+	} else if inputFlags == 1 {
+		if !opts.IO.CanPrompt() && opts.InputType == InputTypeEditor {
+			return &cmdutil.FlagError{Err: errors.New("--body or --web required when not running interactively")}
+		}
+	} else if inputFlags > 1 {
+		return &cmdutil.FlagError{Err: fmt.Errorf("specify only one of --body, --editor, or --web")}
+	}
+
+	return nil
+}
+
+func CommentableRun(opts *CommentableOptions) error {
+	commentable, repo, err := opts.RetrieveCommentable()
+	if err != nil {
+		return err
+	}
+
+	switch opts.InputType {
+	case InputTypeWeb:
+		openURL := commentable.Link() + "#issuecomment-new"
+		if opts.IO.IsStdoutTTY() {
+			fmt.Fprintf(opts.IO.ErrOut, "Opening %s in your browser.\n", utils.DisplayURL(openURL))
+		}
+		return opts.OpenInBrowser(openURL)
+	case InputTypeEditor:
+		var body string
+		if opts.Interactive {
+			body, err = opts.InteractiveEditSurvey()
+		} else {
+			body, err = opts.EditSurvey()
+		}
+		if err != nil {
+			return err
+		}
+		opts.Body = body
+	}
+
+	if opts.Interactive {
+		cont, err := opts.ConfirmSubmitSurvey()
+		if err != nil {
+			return err
+		}
+		if !cont {
+			return fmt.Errorf("Discarding...")
+		}
+	}
+
+	httpClient, err := opts.HttpClient()
+	if err != nil {
+		return err
+	}
+	apiClient := api.NewClientFromHTTP(httpClient)
+	params := api.CommentCreateInput{Body: opts.Body, SubjectId: commentable.Identifier()}
+	url, err := api.CommentCreate(apiClient, repo.RepoHost(), params)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintln(opts.IO.Out, url)
+	return nil
+}
+
+func CommentableConfirmSubmitSurvey() (bool, error) {
+	var confirm bool
+	submit := &survey.Confirm{
+		Message: "Submit?",
+		Default: true,
+	}
+	err := survey.AskOne(submit, &confirm)
+	return confirm, err
+}
+
+func CommentableInteractiveEditSurvey(cf func() (config.Config, error), io *iostreams.IOStreams) func() (string, error) {
+	return func() (string, error) {
+		var body string
+		editorCommand, err := cmdutil.DetermineEditor(cf)
+		if err != nil {
+			return "", err
+		}
+		prompt := &surveyext.GhEditor{
+			EditorCommand: editorCommand,
+			Editor: &survey.Editor{
+				Message:  "Body",
+				FileName: "*.md",
+			},
+		}
+		err = survey.AskOne(prompt, &body)
```

Comment:
I know that I might sound repetitive by re-raising the same issue as before, but I must admit that I still do not understand the purpose of this prompt. I'm being asked this:

```
? Body [(e) to launch vim]
```

My options are: type <kbd>e</kbd> to launch my text editor, or press <kbd>Ctrl-C</kbd> to bail out. Now, I'm bringing this up because I'm not sure what value this prompt brings to the commenting flow? When this prompt appears during `gh issue/pr create`, it makes sense to ask to open the editor because the user can very well choose to skip it since it's optional. Skipping writing a body makes no sense for an action (commenting) whose only input is a written body.

Couldn't we just open an editor? If the user wants to bail out at that point, they can just close their editor.

Reply:
I felt the prompt was needed here since jumping into the editor straight from a command being run would be jarring. This doesn't necessarily need to be a `survey` prompt, I just wanted to give users an opportunity to knowingly jump into their editor. Not sure what technically would make sense here but it could simply be:

```
$ gh pr comment 123
Press **enter** to draft your comment in {editor}
```

Reply:
Thank you for explaining. If the purpose is to provide a keypress guard before opening a text editor program, then the prompt would make sense, and I guess it wouldn't matter much whether the specific key was <kbd>e</kbd> or <kbd>Enter</kbd>.

With my comment, I'm just amplifying the perspective of users who would prefer if the editor was launched right away when we know that the user will want to write multi-line text: https://github.com/cli/cli/issues/390 https://github.com/cli/cli/issues/1090. I personally would also prefer that we skip a prompt. But of course, I'll leave the decision up to you.